src/services/suggestionDiagnostics.ts

```diff
@@ -109,7 +113,118 @@ namespace ts {
         }
     }
 
+    function addConvertToAsyncFunctionDiagnostics(node: Node, checker: TypeChecker, diags: DiagnosticWithLocation[]): void {
+        if (isAsyncFunction(node)) {
+            return;
+        }
+
+        if (isFunctionLikeDeclaration(node) && !node.body) {
+            return; // break on ambient functions
+        }
+
+        const returnType = checker.getTypeAtLocation(node);
+        if (!returnType) {
+            return;
+        }
+
+        if (checker.isPromiseLikeType(returnType)) {
+            // collect all the return statements
+            // check that a property access expression exists in there and that it is a handler
+            let varDeclFlagsMap: Map<NodeFlags | undefined> = new MapCtr();
+            const retStmts = getReturnStatementsWithPromiseCallbacks(node, varDeclFlagsMap);
+            if (retStmts.length > 0) {
+                diags.push(createDiagnosticForNode(isVariableDeclaration(node.parent) ? node.parent.name : node, Diagnostics.This_may_be_converted_to_an_async_function));
+            }
+        }
+    }
+
     function getErrorNodeFromCommonJsIndicator(commonJsModuleIndicator: Node): Node {
         return isBinaryExpression(commonJsModuleIndicator) ? commonJsModuleIndicator.left : commonJsModuleIndicator;
     }
+
+    export function getReturnStatementsWithPromiseCallbacks(node: Node, varDeclFlagsMap: Map<NodeFlags | undefined>): [Node[], Map<NodeFlags | undefined>, ReturnStatement | undefined] {
+        const retStmts: Node[] = [];
+        let hasFollowingRetStmt: ReturnStatement | undefined;
+        forEachChild(node, visit);
+
+        function visit(child: Node) {
+
+            if (isFunctionLike(child)) {
+                return;
+            }
+
+            if (isReturnStatement(child)) {
+                forEachChild(child, hasCallback);
+            }
+
+            function hasCallback(returnChild: Node) {
+                //const symbol = checker.getSymbolAtLocation(returnChild);
+
+                if (isCallback(returnChild)) {
+                    retStmts.push(child as ReturnStatement);
+                }
+                else if (isIdentifier(returnChild) && isReturnStatement(child)
+                    && child.expression && isIdentifier(child.expression)) {

```

Comment:
i am a bit confused about this condition.. so are we checking that the expression of the return statement is an identifier?